### PR TITLE
Materialization API Updates + UI 

### PR DESCRIPTION
**Pull request by shangyian**
#### API Changes

From the API perspective, this PR aims to turn the existing materialization API endpoints into more modularized actions, inferring config where possible rather than relying on user-supplied config. The new user flow for configuring materialization looks like this:
1. Set partition columns (if any) on the node, either temporal or categorical. For Druid materializations, temporal partitions are required.
```bash
curl -X POST 'http://localhost:8000/nodes/default.hard_hat/columns/hire_date/partition'
-d '{
    "type_": "temporal",
    "expression": "",
}'
```

2. Configure a materialization for the node. This involves selecting the engine, the schedule, and providing some optional config (for example, Spark job configs) if needed. This configuration can be edited if needed.
```bash
curl -X POST 'http://localhost:8000/nodes/default.hard_hat/materialization'
-d '{
    "engine": {
        "name": "spark",
        "version": "3.4",
    },
    "config": {"spark": {"spark.executor.memory": "7g"}},
    "schedule": "0 * * * *",
}'
```

3. Kick off a backfill for the configured materialization. If a temporal partition is set on the node, this involves providing a range of values for that partition.
```bash
curl -X POST 'http://localhost:8000/nodes/default.hard_hat/materializations/hire_date_spark/backfill'
-d '{
    "column_name": "hire_date",
    "range": ["20230101", "20230201"],
}'
```

#### UI Support for Configuring Materializations and Backfills

This PR also adds UI support for the above actions.

1. Setting column partitions:

https://github.com/DataJunction/dj/assets/9524628/733b5642-d8c2-4232-adbc-c539e9b3bd3d

2. Configuring materialization:

https://github.com/DataJunction/dj/assets/9524628/999e6baf-995c-44fb-bea6-c1128cf4c425

3. Run backfill:

https://github.com/DataJunction/dj/assets/9524628/c2cc1641-b9c0-41f4-873e-72b93804c2ac

4. After posting availability state with:
```
curl -X POST 'http://localhost:8000/data/default.repair_orders_cube/availability/' \ 
-H 'Content-Type: application/json' \
-d '{
    "catalog": "druid",
    "schema": "",
    "table": "dj_repair_orders_cube__v1",
    "valid_through_ts": "20230902",
    "temporal_partitions": ["20230901", "20230902"],
    "min_temporal_partition": ["20230901"],
    "max_temporal_partition": ["20230902"]
}'
```

The UI displays:

Screenshot 2023-10-09 at 4 50 44 PM

### Test Plan

Running DJ locally

- [ ] PR has an associated issue: #809 
- [x] `make check` passes
- [x] `make test` shows 100% unit test coverage

### CODE REVIEW

File: datajunction-server/datajunction_server/materialization/jobs/cube_materialization.py [hunk 1]
```diff
@@ -82,46 +87,42 @@ def build_druid_spec(self, cube_config: DruidCubeConfig, node_name: str) -> Dict
...
-            if cube_config.partitions
-            else []
-        )
-        if not temporal_partitions:
-            raise DJException(
-                "Druid ingestion requires a temporal partition to be specified",
+        granularity, timestamp_column = None, None
+        for col in node_revision.columns:
+            if col.partition and col.partition.type_ == PartitionType.TEMPORAL:
+                granularity = "DAY"
```
Comment (samredai) [inline datajunction-server/datajunction_server/materialization/jobs/cube_materialization.py:93]: Does this mean temporal partitions on cubes are always daily?

**Review by samredai — APPROVED** [review]
LGTM, the UI changes are super exciting. I just left a tiny question about one line.

**Review by agorajek — APPROVED** [review]
Big PR... awesome features. Not sure if you had a chance to test the partition expression end-2-end but either way, I think its much more rare case and we can test it out "in Prod". Thanks @shangyian !

File: datajunction-server/datajunction_server/materialization/jobs/cube_materialization.py (hunk 1 above)
Comment (shangyian) [reply datajunction-server/datajunction_server/materialization/jobs/cube_materialization.py]: Good point -- I just updated this to add support for users providing a datetime format (like `yyyyMMdd`) + granularity and DJ building out the expected expression/intervals based on that.

**Pull request merged**